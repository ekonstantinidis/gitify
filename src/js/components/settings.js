import React from 'react';
import Toggle from 'react-toggle';

const ipcRenderer = window.require('electron').ipcRenderer;

var Actions = require('../actions/actions');
var SettingsStore = require('../stores/settings');

export default class SettingsPage extends React.Component {

  constructor(props) {
    super(props);

    var settings = SettingsStore.getSettings();
    this.state = {
      participating: settings.participating,
      playSound: settings.playSound,
      showNotifications: settings.showNotifications,
      markOnClick: settings.markOnClick,
      openAtStartup: settings.openAtStartup
    };
  }

  toggleSetting(key, event) {
    Actions.setSetting(key, event.target.checked);
  }

  checkForUpdates() {
    ipcRenderer.send('check-update');
  }

  appQuit() {
    ipcRenderer.send('app-quit');
  }

  render() {
    var pjson = require('../../../package.json');
    var version = pjson.version;

    return (
      <div className="container-fluid main-container settings">
        <div className="row">
          <div className="col-xs-8">Show only participating</div>
          <div className="col-xs-4">
            <Toggle
              defaultChecked={this.state.participating}
              onChange={this.toggleSetting.bind(this, 'participating')} />
          </div>
        </div>
        <div className="row">
          <div className="col-xs-8">Play sound</div>
          <div className="col-xs-4">
            <Toggle
              defaultChecked={this.state.playSound}
              onChange={this.toggleSetting.bind(this, 'playSound')} />
          </div>
        </div>
        <div className="row">
          <div className="col-xs-8">Show notifications</div>
          <div className="col-xs-4">
            <Toggle
              defaultChecked={this.state.showNotifications}
              onChange={this.toggleSetting.bind(this, 'showNotifications')} />
          </div>
        </div>

        <div className="row">
          <div className="col-xs-8">On Click, Mark as Read</div>
          <div className="col-xs-4">
            <Toggle
              defaultChecked={this.state.markOnClick}
              onChange={this.toggleSetting.bind(this, 'markOnClick')} />
          </div>
        </div>

        <div className="row">
          <div className="col-xs-8">Open at startup</div>
          <div className="col-xs-4">
            <Toggle
              defaultChecked={this.state.openAtStartup}
              onChange={this.toggleSetting.bind(this, 'openAtStartup')} />
          </div>
        </div>
        <div className="row">
          <div className="col-xs-6">
            <button
              className="btn btn-block btn-primary btn-close"
              onClick={this.checkForUpdates}>
              <i className="fa fa-cloud-download" />
              Update
            </button>
          </div>
          <div className="col-xs-6">
            <button
              className="btn btn-block btn-danger btn-close"
              onClick={this.appQuit}>
              <i className="fa fa-power-off" />
              Quit Gitify
            </button>
          </div>
        </div>

        <div className="row footer">
          <div className="col-xs-12 text-right">Gitify - Version: {version}</div>
        </div>
      </div>
    );
  }
};
