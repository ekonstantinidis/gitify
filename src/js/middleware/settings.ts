import { UPDATE_SETTING } from '../actions';
import { setAppearance } from '../utils/appearance';
import { setAutoLaunch } from '../utils/comms';

export default () => (next) => (action) => {
  switch (action.type) {
    case UPDATE_SETTING:
      if (action.setting === 'openAtStartup') {
        setAutoLaunch(action.value);
      }

      if (action.setting === 'appearance') {
        setAppearance(action.value);
      }
  }

  return next(action);
};
