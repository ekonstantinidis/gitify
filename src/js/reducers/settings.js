import { UPDATE_SETTING, HAS_STARRED, TOGGLE_SETTINGS_MODAL } from '../actions';
import { Map } from 'immutable';

const initialState = Map({
  participating: false,
  playSound: true,
  showNotifications: true,
  markOnClick: false,
  openAtStartup: false,
  showSettingsModal: false,
  hasStarred: false,
  showAppIcon: 'both',
  isEnterprise: false,
  baseUrl: 'github.com',
  clientId: '3fef4433a29c6ad8f22c',
  clientSecret: '9670de733096c15322183ff17ed0fc8704050379'
});

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case UPDATE_SETTING:
      return state.set(action.setting, action.value);
    case HAS_STARRED.SUCCESS:
      return state.set('hasStarred', true);
    case HAS_STARRED.FAILURE:
      return state.set('hasStarred', false);
    case TOGGLE_SETTINGS_MODAL:
      return state.set('showSettingsModal', !state.get('showSettingsModal'));
    default:
      return state;
  }
};
