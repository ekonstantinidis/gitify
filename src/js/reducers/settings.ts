import { UPDATE_SETTING } from '../actions';
import { SettingsState } from '../../types/reducers';
import { Appearance } from '../../types';

const initialState: SettingsState = {
  participating: false,
  playSound: true,
  showNotifications: true,
  markOnClick: false,
  openAtStartup: false,
  appearance: Appearance.SYSTEM,
};

export default function reducer(state = initialState, action): SettingsState {
  switch (action.type) {
    case UPDATE_SETTING:
      return { ...state, [action.setting]: action.value };
    default:
      return state;
  }
}
