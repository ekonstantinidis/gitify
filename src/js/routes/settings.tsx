const { ipcRenderer, remote } = require('electron');

import * as React from 'react';
import { connect } from 'react-redux';
import { ArrowLeftIcon } from '@primer/octicons-react';

import { AppState, SettingsState } from '../../types/reducers';
import { fetchNotifications, updateSetting, logout } from '../actions';
import { FieldCheckbox } from '../components/ui/checkbox';
import { updateTrayIcon } from '../utils/comms';

const isLinux = remote.process.platform === 'linux';

interface IProps {
  hasMultipleAccounts: boolean;
  fetchNotifications: () => any;
  logout: () => any;
  settings: SettingsState;
  updateSetting: any;
  history: any;
}

export class SettingsRoute extends React.Component<IProps> {
  constructor(props) {
    super(props);

    this.state = {
      participating: props.settings.participating,
    };
  }

  static getDerivedStateFromProps(props, state) {
    if (props.settings.participating !== state.participating) {
      props.fetchNotifications();
    }

    return {
      participating: props.settings.participating,
    };
  }

  logout() {
    this.props.logout();
    this.props.history.goBack();
    updateTrayIcon();
  }

  quitApp() {
    ipcRenderer.send('app-quit');
  }

  goToEnterprise() {
    return this.props.history.replace('/enterpriselogin');
  }

  render() {
    const { hasMultipleAccounts, settings } = this.props;

    const footerButtonClass =
      'bg-gray-400 hover:bg-gray-500 hover:text-white rounded py-1 px-2 my-1 mx-2 text-sm focus:outline-none';

    return (
      <div className="flex flex-1 flex-col">
        <div className="flex justify-between items-center mt-4 py-2 mx-8">
          <button
            className="focus:outline-none"
            aria-label="Go Back"
            onClick={() => this.props.history.goBack()}
          >
            <ArrowLeftIcon size={20} className="hover:text-gray-400" />
          </button>

          <h3 className="text-lg font-semibold">Settings</h3>
        </div>

        <div className="flex-1 px-8">
          <FieldCheckbox
            name="showOnlyParticipating"
            label="Show only participating"
            checked={settings.participating}
            onChange={(evt) =>
              this.props.updateSetting('participating', evt.target.checked)
            }
          />
          <FieldCheckbox
            name="playSound"
            label="Play sound"
            checked={settings.playSound}
            onChange={(evt) =>
              this.props.updateSetting('playSound', evt.target.checked)
            }
          />
          <FieldCheckbox
            name="showNotifications"
            label="Show notifications"
            checked={settings.showNotifications}
            onChange={(evt) =>
              this.props.updateSetting('showNotifications', evt.target.checked)
            }
          />
          <FieldCheckbox
            name="onClickMarkAsRead"
            label="On Click, Mark as Read"
            checked={settings.markOnClick}
            onChange={(evt) =>
              this.props.updateSetting('markOnClick', evt.target.checked)
            }
          />
          {!isLinux && (
            <FieldCheckbox
              name="openAtStartUp"
              label="Open at startup"
              checked={settings.openAtStartup}
              onChange={(evt) =>
                this.props.updateSetting('openAtStartup', evt.target.checked)
              }
            />
          )}
        </div>

        <div className="flex justify-between items-center bg-gray-300 py-4 px-8">
          <small className="font-semibold">
            Gitify v{remote.app.getVersion()}
          </small>

          <div>
            <button
              className={footerButtonClass}
              aria-label="Login with GitHub Enterprise"
              onClick={this.goToEnterprise.bind(this)}
            >
              Add Enterprise
            </button>

            <button
              className={footerButtonClass}
              aria-label="Logout"
              onClick={this.logout.bind(this)}
            >
              {hasMultipleAccounts ? 'Logout from all accounts' : 'Logout'}
            </button>

            <button
              className={`${footerButtonClass} mr-0`}
              aria-label="Quit Gitify"
              onClick={this.quitApp}
            >
              Quit
            </button>
          </div>
        </div>
      </div>
    );
  }
}

export function mapStateToProps(state: AppState) {
  const enterpriseAccounts = state.auth.enterpriseAccounts;
  const isGitHubLoggedIn = state.auth.token !== null;
  const hasMultipleAccounts = isGitHubLoggedIn
    ? enterpriseAccounts.length > 0
    : enterpriseAccounts.length > 1;

  return {
    settings: state.settings,
    hasMultipleAccounts,
  };
}

export default connect(mapStateToProps, {
  updateSetting,
  fetchNotifications,
  logout,
})(SettingsRoute);
