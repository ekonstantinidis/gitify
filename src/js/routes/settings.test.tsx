import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

const { ipcRenderer } = require('electron');

import { SettingsRoute, mapStateToProps } from './settings';
import {
  SettingsState,
  AppState,
  AuthState,
  EnterpriseAccount,
} from '../../types/reducers';

describe('routes/settings.tsx', () => {
  const props = {
    updateSetting: jest.fn(),
    fetchNotifications: jest.fn(),
    hasMultipleAccounts: false,
    logout: jest.fn(),
    history: {
      push: jest.fn(),
      goBack: jest.fn(),
      replace: jest.fn(),
    },
    settings: {
      participating: false,
      playSound: true,
      showNotifications: true,
      markOnClick: false,
      openAtStartup: false,
    } as SettingsState,
  };

  beforeEach(function () {
    spyOn(ipcRenderer, 'send');

    props.updateSetting.mockReset();
    props.history.goBack.mockReset();
    props.history.push.mockReset();
    props.history.replace.mockReset();
  });

  describe('mapStateToProps', () => {
    const state = {
      auth: {
        token: '123-456',
        enterpriseAccounts: [{} as EnterpriseAccount],
      } as AuthState,
      settings: {
        participating: false,
      } as SettingsState,
    } as AppState;
    it('should test the method', () => {
      const mappedProps = mapStateToProps(state);

      expect(mappedProps.hasMultipleAccounts).toBeTruthy();
      expect(mappedProps.settings.participating).toBeFalsy();
    });

    it('should recognize when only one account logged in', () => {
      const mappedProps = mapStateToProps({
        ...state,
        auth: {
          ...state.auth,
          token: null,
        },
      });

      expect(mappedProps.hasMultipleAccounts).toBeFalsy();
    });
  });

  it('should render itself & its children', () => {
    const tree = TestRenderer.create(
      <MemoryRouter>
        <SettingsRoute {...props} />
      </MemoryRouter>
    );
    expect(tree).toMatchSnapshot();
  });

  it('should press the logout', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('Logout'));

    expect(props.logout).toHaveBeenCalledTimes(1);

    expect(ipcRenderer.send).toHaveBeenCalledTimes(1);
    expect(ipcRenderer.send).toHaveBeenCalledWith('update-icon');
    expect(props.history.goBack).toHaveBeenCalledTimes(1);
  });

  it('should call the componentWillReceiveProps method', function () {
    const { rerender } = render(<SettingsRoute {...props} />);

    expect(props.fetchNotifications).toHaveBeenCalledTimes(0);

    const updatedSettings = {
      ...props.settings,
      participating: !props.settings.participating,
    };

    rerender(<SettingsRoute {...props} settings={updatedSettings} />);

    expect(props.fetchNotifications).toHaveBeenCalledTimes(1);
  });

  it('should go back by pressing the icon', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);
    fireEvent.click(getByLabelText('Close Settings'));
    expect(props.history.goBack).toHaveBeenCalledTimes(1);
  });

  it('should toggle the showOnlyParticipating checkbox', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('Show only participating'), {
      target: { checked: true },
    });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the playSound checkbox', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('Play sound'), {
      target: { checked: true },
    });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the showNotifications checkbox', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('Show notifications'), {
      target: { checked: true },
    });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the onClickMarkAsRead checkbox', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('On Click, Mark as Read'), {
      target: { checked: true },
    });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the openAtStartup checkbox', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('Open at startup'), {
      target: { checked: true },
    });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should go to the enterprise login route', () => {
    const { getByLabelText } = render(
      <MemoryRouter>
        <SettingsRoute {...props} />
      </MemoryRouter>
    );
    fireEvent.click(getByLabelText('Login with GitHub Enterprise'));
    expect(props.history.replace).toHaveBeenCalledWith('/enterpriselogin');
  });

  it('should quit the app', () => {
    const { getByLabelText } = render(
      <MemoryRouter>
        <SettingsRoute {...props} />
      </MemoryRouter>
    );
    fireEvent.click(getByLabelText('Quit Gitify'));
    expect(ipcRenderer.send).toHaveBeenCalledWith('app-quit');
  });
});
