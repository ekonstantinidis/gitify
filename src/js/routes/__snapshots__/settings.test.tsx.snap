// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes/settings.tsx should render itself & its children 1`] = `
<div
  className="flex flex-1 flex-col"
>
  <div
    className="flex justify-between items-center mt-4 py-2 mx-8"
  >
    <button
      aria-label="Go Back"
      className="focus:outline-none"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="hover:text-gray-400"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path fill-rule=\\"evenodd\\" d=\\"M7.78 12.53a.75.75 0 01-1.06 0L2.47 8.28a.75.75 0 010-1.06l4.25-4.25a.75.75 0 011.06 1.06L4.81 7h7.44a.75.75 0 010 1.5H4.81l2.97 2.97a.75.75 0 010 1.06z\\"></path>",
          }
        }
        fill="currentColor"
        height={20}
        role="img"
        style={
          Object {
            "display": "inline-block",
            "userSelect": "none",
            "verticalAlign": "text-bottom",
          }
        }
        viewBox="0 0 16 16"
        width={20}
      />
    </button>
    <h3
      className="text-lg font-semibold"
    >
      Settings
    </h3>
  </div>
  <div
    className="flex-1 px-8"
  >
    <div
      className="mt-1 mb-2"
    >
      <label
        className="inline-flex items-center mt-2"
      >
        <input
          checked={false}
          className="form-checkbox h-5 w-5 text-red-600"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="ml-2 text-gray-700"
        >
          Show only participating
        </span>
      </label>
    </div>
    <div
      className="mt-1 mb-2"
    >
      <label
        className="inline-flex items-center mt-2"
      >
        <input
          checked={true}
          className="form-checkbox h-5 w-5 text-red-600"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="ml-2 text-gray-700"
        >
          Play sound
        </span>
      </label>
    </div>
    <div
      className="mt-1 mb-2"
    >
      <label
        className="inline-flex items-center mt-2"
      >
        <input
          checked={true}
          className="form-checkbox h-5 w-5 text-red-600"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="ml-2 text-gray-700"
        >
          Show notifications
        </span>
      </label>
    </div>
    <div
      className="mt-1 mb-2"
    >
      <label
        className="inline-flex items-center mt-2"
      >
        <input
          checked={false}
          className="form-checkbox h-5 w-5 text-red-600"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="ml-2 text-gray-700"
        >
          On Click, Mark as Read
        </span>
      </label>
    </div>
    <div
      className="mt-1 mb-2"
    >
      <label
        className="inline-flex items-center mt-2"
      >
        <input
          checked={false}
          className="form-checkbox h-5 w-5 text-red-600"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="ml-2 text-gray-700"
        >
          Open at startup
        </span>
      </label>
    </div>
  </div>
  <div
    className="flex justify-between items-center bg-gray-300 py-4 px-8"
  >
    <small
      className="font-semibold"
    >
      Gitify v
      0.0.1
    </small>
    <div>
      <button
        aria-label="Login with GitHub Enterprise"
        className="hover:text-gray-500 py-1 px-2 my-1 mx-2 focus:outline-none"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          aria-labelledby="iconAddAccountId"
          className="w-5 h-5"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="iconAddAccountId"
          >
            Add Account
          </title>
          <g>
            <path
              d="M640 224v32a16 16 0 01-16 16h-64v64a16 16 0 01-16 16h-32a16 16 0 01-16-16v-64h-64a16 16 0 01-16-16v-32a16 16 0 0116-16h64v-64a16 16 0 0116-16h32a16 16 0 0116 16v64h64a16 16 0 0116 16z"
              fill="currentColor"
              opacity={0.4}
            />
            <path
              d="M224 256A128 128 0 1096 128a128 128 0 00128 128zm89.6 32h-16.7a174.08 174.08 0 01-145.8 0h-16.7A134.43 134.43 0 000 422.4V464a48 48 0 0048 48h352a48 48 0 0048-48v-41.6A134.43 134.43 0 00313.6 288z"
              fill="currentColor"
            />
          </g>
        </svg>
      </button>
      <button
        aria-label="Logout"
        className="hover:text-gray-500 py-1 px-2 my-1 mx-2 focus:outline-none"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          aria-labelledby="iconLogoutId"
          className="w-5 h-5"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="iconLogoutId"
          >
            Logout from all accounts
          </title>
          <g>
            <path
              d="M180 448H96a96 96 0 01-96-96V160a96 96 0 0196-96h84a12 12 0 0112 12v40a12 12 0 01-12 12H96a32 32 0 00-32 32v192a32 32 0 0032 32h84a12 12 0 0112 12v40a12 12 0 01-12 12z"
              fill="currentColor"
              opacity={0.4}
            />
            <path
              d="M353 88.3l151.9 150.6a24 24 0 010 34.1l-152 150.8a24.08 24.08 0 01-33.9-.1l-21.9-21.9a24.07 24.07 0 01.8-34.7l77.6-71.1H184a23.94 23.94 0 01-24-24v-32a23.94 23.94 0 0124-24h191.5l-77.6-71.1a24 24 0 01-.7-34.6l21.9-21.9a24 24 0 0133.9-.1z"
              fill="currentColor"
            />
          </g>
        </svg>
      </button>
      <button
        aria-label="Quit Gitify"
        className="hover:text-gray-500 py-1 px-2 my-1 mx-2 focus:outline-none mr-0"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          aria-labelledby="iconQuitId"
          className="w-5 h-5"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="iconQuitId"
          >
            Quit Application
          </title>
          <g>
            <path
              d="M272 0a23.94 23.94 0 0124 24v240a23.94 23.94 0 01-24 24h-32a23.94 23.94 0 01-24-24V24a23.94 23.94 0 0124-24z"
              fill="currentColor"
              opacity={0.4}
            />
            <path
              d="M504 256c0 136.8-110.8 247.7-247.5 248C120 504.3 8.2 393 8 256.4A248 248 0 01111.8 54.2a24.07 24.07 0 0135 7.7L162.6 90a24 24 0 01-6.6 31 168 168 0 00100 303c91.6 0 168.6-74.2 168-169.1a168.07 168.07 0 00-68.1-134 23.86 23.86 0 01-6.5-30.9l15.8-28.1a24 24 0 0134.8-7.8A247.51 247.51 0 01504 256z"
              fill="currentColor"
            />
          </g>
        </svg>
      </button>
    </div>
  </div>
</div>
`;
